library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java

```diff
@@ -17,11 +17,22 @@ public TarTruncator(DataWriter dataWriter) {
     public DownloadTask.State transferData(DownloadTask.State state, InputStream in) throws StopRequestException {
         DownloadTask.State newState = state;
         try {
-            byte[] buffer = new byte[Constants.BUFFER_SIZE];
-            int readLast;
-            while ((readLast = readBuffer(in, buffer)) > 0 && isNotFullOfZeroes(buffer)) {
-                newState = dataWriter.write(newState, buffer, readLast);
+            byte[] buffer = new byte[Constants.TAR_RECORD_SIZE];
+            byte[] buffer2 = new byte[Constants.TAR_RECORD_SIZE];
+            byte[] swappingRef;
+            int read;
+            int readLast = 0;
+
+            while ((read = readRecord(in, buffer)) > 0) {
+                newState = dataWriter.write(newState, buffer2, readLast);
+                swappingRef = buffer2;
+                buffer2 = buffer;
+                buffer = swappingRef;
+                readLast = read;
```

Comment:
This is a bit hard to parse atm. Get it eventually, but some renaming would help a bit